<a href="https://colab.research.google.com/github/Amit-Nair04/Data-Analytics-project/blob/main/Data%20Analytics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#TASK 1

In [52]:
#Importing Pandas
import pandas as pd

In [53]:
#Uploading the CSV file
from google.colab import files

uploaded = files.upload()

Saving bangalore_tech_salaries.csv to bangalore_tech_salaries (1).csv


In [54]:
#Reading the CSV file
df = pd.read_csv("bangalore_tech_salaries.csv")

In [55]:
#Displaying the first 5 rows
df.head()

,Employee_ID,Role,years_exp,Current_CTC,Previous_CTC,Company,company_TYPE,Skills,Location,Education_Tier,Joining_Year,Work_Mode
0,BLR0065,DS,6,49.4,32.4 LPA,Ola,Unicorn,"JIRA, JavaScript, Spring Boot, Figma, NumPy, A...",HSR Layout,Tier-1,2022,Remote
1,BLR0080,DevOps,0,9.7,NaN,Walmart Labs,MNC,"Tableau, Python, Figma, JavaScript, TensorFlow",Electronic City,2,2024,Remote
2,BLR0166,SDE Backend,6,"3,360,000",24.2 LPA,Postman,Mid-size,"Agile, Excel, GCP, MongoDB, Java, ML",Whitefield,Tier-1,2023,Hybrid
3,BLR0219,SDE Frontend,8,41.4 LPA,30.3,Amazon India,MNC,"PyTorch, Redis, System Design",MG Road,Tier-2,2022,Work from Office
4,BLR0491,Product Lead,5,"3,640,000",30.6 LPA,QuickPay AI,early-stage,"Tableau, Deep Learning, Agile",Whitefield,Tier-2,2023,Onsite


In [56]:
#Finding the shape
df.shape

(1015, 12)

In [57]:
#checking information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1015 entries, 0 to 1014
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Employee_ID     1015 non-null   object
 1   Role            1015 non-null   object
 2   years_exp       1015 non-null   int64 
 3   Current_CTC     1015 non-null   object
 4   Previous_CTC    815 non-null    object
 5   Company         1015 non-null   object
 6   company_TYPE    1015 non-null   object
 7   Skills          988 non-null    object
 8   Location        995 non-null    object
 9   Education_Tier  1015 non-null   object
 10  Joining_Year    1015 non-null   int64 
 11  Work_Mode       1015 non-null   object
dtypes: int64(2), object(10)
memory usage: 95.3+ KB


In [58]:
#Checking the data types
df.dtypes

,0
Employee_ID,object
Role,object
years_exp,int64
Current_CTC,object
Previous_CTC,object
Company,object
company_TYPE,object
Skills,object
Location,object
Education_Tier,object


In [59]:
#counting missing values
df.isnull().sum()

,0
Employee_ID,0
Role,0
years_exp,0
Current_CTC,0
Previous_CTC,200
Company,0
company_TYPE,0
Skills,27
Location,20
Education_Tier,0


In [60]:
#counting dulicate rows
df.duplicated().sum()

np.int64(15)

#TASK 2

In [61]:
#Renaming all cols to snake_case
df.columns = df.columns.str.strip().str.lower()

df.columns

Index(['employee_id', 'role', 'years_exp', 'current_ctc', 'previous_ctc',
       'company', 'company_type', 'skills', 'location', 'education_tier',
       'joining_year', 'work_mode'],
      dtype='object')

In [62]:
# Checking  all role variants
df["role"].unique()

array(['DS', 'DevOps', 'SDE Backend', 'SDE Frontend', 'Product Lead',
       'Product Manager', 'Site Reliability Engineer', 'DA', 'UI/UX',
       'SDE-Frontend', 'UI Designer', 'ML Engineer', 'Frontend Dev',
       'Data Scientist', 'Full Stack Engineer', 'Designer',
       'Product Designer', 'BA', 'Frontend Engineer', 'BI Analyst',
       'Backend Dev', 'Sr PM', 'FE', 'SDE-Backend', 'Data Analyst',
       'Business Analyst', 'Backend Developer',
       'Business Systems Analyst', 'Data Science Engineer', 'FullStack',
       'BE', 'Frontend Developer', 'Backend Engineer', 'UX Designer',
       'Infra Engineer', 'PM', 'SDE FS', 'DevOps Engineer',
       'Analytics Engineer', 'Fullstack Dev', 'SDE Full-Stack', 'SRE'],
      dtype=object)

In [63]:
# Standardizing  role names (AI Assisted)
role_map = {
    'DS': 'Data Scientist',
    'Data Scientist': 'Data Scientist',
    'Data Science Engineer': 'Data Scientist',
    'ML Engineer': 'ML Engineer',

    'SDE Backend': 'SDE Backend',
    'SDE-Backend': 'SDE Backend',
    'Backend Dev': 'SDE Backend',
    'Backend Developer': 'SDE Backend',
    'Backend Engineer': 'SDE Backend',
    'BE': 'SDE Backend',

    'SDE Frontend': 'SDE Frontend',
    'SDE-Frontend': 'SDE Frontend',
    'Frontend Dev': 'SDE Frontend',
    'Frontend Developer': 'SDE Frontend',
    'Frontend Engineer': 'SDE Frontend',

    'Full Stack Engineer': 'SDE Full-Stack',
    'FullStack': 'SDE Full-Stack',
    'Fullstack Dev': 'SDE Full-Stack',
    'SDE FS': 'SDE Full-Stack',
    'SDE Full-Stack': 'SDE Full-Stack',

    'DevOps': 'DevOps Engineer',
    'DevOps Engineer': 'DevOps Engineer',
    'Site Reliability Engineer': 'DevOps Engineer',
    'SRE': 'DevOps Engineer',

    'Product Manager': 'Product Manager',
    'PM': 'Product Manager',
    'Product Lead': 'Product Manager',

    'UI Designer': 'UI/UX Designer',
    'UX Designer': 'UI/UX Designer',
    'UI/UX': 'UI/UX Designer',
    'Product Designer': 'UI/UX Designer',
    'Designer': 'UI/UX Designer',

    'Data Analyst': 'Data Analyst',
    'DA': 'Data Analyst',
    'BI Analyst': 'Data Analyst',
    'Analytics Engineer': 'Data Analyst',

    'Business Analyst': 'Business Analyst',
    'Business Systems Analyst': 'Business Analyst',

    'Infra Engineer': 'Infrastructure Engineer',
    'FE': 'SDE Frontend',
    'Sr PM': 'Product Manager'
}

df["role_clean"] = df["role"].replace(role_map)

In [64]:
df["role_clean"].value_counts()

,count
role_clean,
SDE Backend,116
UI/UX Designer,116
Product Manager,113
DevOps Engineer,107
SDE Frontend,105
Data Analyst,101
SDE Full-Stack,97
Data Scientist,96
Business Analyst,69


In [65]:
df["role_clean"] = df["role_clean"].replace({"BA":"Business Analyst"})

In [66]:
df["role_clean"].value_counts()

,count
role_clean,
SDE Backend,116
UI/UX Designer,116
Product Manager,113
Business Analyst,112
DevOps Engineer,107
SDE Frontend,105
Data Analyst,101
SDE Full-Stack,97
Data Scientist,96


In [67]:
df["company_type"].unique()

array(['Unicorn', 'MNC', 'Mid-size', 'early-stage', 'Early-stage',
       'EARLY-STAGE', 'mnc', 'unicorn', 'MID-SIZE', 'mid-size', 'UNICORN'],
      dtype=object)

In [68]:
company_map = {
    "mnc": "MNC",
    "MID-SIZE": "Mid-size",
    "mid-size": "Mid-size",
    "early-stage": "Early-stage",
    "EARLY-STAGE": "Early-stage",
    "unicorn": "Unicorn",
    "UNICORN": "Unicorn"
}

df["company_type_clean"] = df["company_type"].replace(company_map)

In [69]:
df["company_type_clean"].value_counts()

,count
company_type_clean,
MNC,348
Mid-size,261
Unicorn,207
Early-stage,199


In [70]:
df["education_tier"].unique()

array(['Tier-1', '2', 'Tier-2', 'Tier 1', 'Tier 2', 'T3', 'Tier-3', 'T2',
       '1', 'T1', 'Tier 3', '3'], dtype=object)

In [71]:
education_map = {
    "Tier-1": "Tier 1",
    "T1": "Tier 1",
    "1": "Tier 1",

    "Tier-2": "Tier 2",
    "T2": "Tier 2",
    "2": "Tier 2",

    "Tier-3": "Tier 3",
    "T3": "Tier 3",
    "3": "Tier 3"
}

df["education_tier_clean"] = df["education_tier"].replace(education_map)

In [72]:
df["education_tier_clean"].value_counts()

,count
education_tier_clean,
Tier 2,510
Tier 3,309
Tier 1,196


In [73]:
# Displaying first 10 Current CTC values
df["current_ctc"].head(10)

,current_ctc
0,49.4
1,9.7
2,"3,360,000"
3,41.4 LPA
4,"3,640,000"
5,32.1
6,24.6
7,21.7 LPA
8,21.0
9,26.9 LPA


In [74]:
# Displaying first 10 Previous CTC values
df["previous_ctc"].head(10)

,previous_ctc
0,32.4 LPA
1,NaN
2,24.2 LPA
3,30.3
4,30.6 LPA
5,22.0 LPA
6,19.3 LPA
7,17.7
8,NaN
9,18.9


In [75]:
# Converting CTC values to LPA (Ai Assisted)
def parse_ctc(x):
    if pd.isna(x):
        return x

    x = str(x).replace("₹", "").replace("LPA", "").replace(",", "").strip()

    value = float(x)

    if value > 100000:
        value = value / 100000

    return value

In [76]:
# Converting  Current CTC and Previous CTC into numeric LPA
df["current_ctc"] = df["current_ctc"].apply(parse_ctc)
df["previous_ctc"] = df["previous_ctc"].apply(parse_ctc)

In [77]:
# Checking  the converted values
df[["current_ctc", "previous_ctc"]].head(10)

,current_ctc,previous_ctc
0,49.4,32.4
1,9.7,NaN
2,33.6,24.2
3,41.4,30.3
4,36.4,30.6
5,32.1,22.0
6,24.6,19.3
7,21.7,17.7
8,21.0,NaN
9,26.9,18.9


In [78]:
# Removing duplicate rows from the dataset
df = df.drop_duplicates()

In [79]:
df.duplicated().sum()

np.int64(0)

In [80]:
# Displaying data types after cleaning
df.dtypes

,0
employee_id,object
role,object
years_exp,int64
current_ctc,float64
previous_ctc,float64
company,object
company_type,object
skills,object
location,object
education_tier,object


In [81]:
# Verifying cleaned role names
df["role_clean"].value_counts()

,count
role_clean,
SDE Backend,114
UI/UX Designer,113
Product Manager,112
Business Analyst,108
DevOps Engineer,107
SDE Frontend,105
Data Analyst,100
SDE Full-Stack,96
Data Scientist,94


In [82]:
# Verifying cleaned company types
df["company_type_clean"].value_counts()

,count
company_type_clean,
MNC,342
Mid-size,257
Unicorn,207
Early-stage,194


In [83]:
# Verifying cleaned education tiers
df["education_tier_clean"].value_counts()

,count
education_tier_clean,
Tier 2,500
Tier 3,306
Tier 1,194


#TASK 3 Q3.1

In [84]:
# Calculating median, mean, minimum and maximum CTC for each role
role_ctc = df.groupby("role_clean")["current_ctc"].agg(
    Median="median",
    Mean="mean",
    Min="min",
    Max="max"
)

role_ctc = role_ctc.sort_values(by="Median", ascending=False)

role_ctc

,Median,Mean,Min,Max
role_clean,,,,
Product Manager,31.30,34.131250,10.8,80.1
ML Engineer,27.50,30.335484,10.0,64.5
Data Scientist,24.15,26.670213,10.9,75.6
Infrastructure Engineer,23.45,22.610000,9.2,45.5
SDE Full-Stack,22.35,25.409375,8.9,71.7
DevOps Engineer,22.10,23.390654,8.8,60.3
SDE Frontend,21.20,22.193333,6.7,84.4
SDE Backend,21.10,23.284210,7.9,55.1
Business Analyst,19.90,21.873148,6.8,52.7


### ANSWER Q3.1

Product Managers have the highest median CTC at 31.30 LPA, while Data Analysts have the lowest median CTC at 16.90 LPA. The higher mean compared to the median for roles like Product Manager indicates that a few highly paid professionals are increasing the average salary.

#TASK 3 Q3.2

In [85]:
# Creating experience bands for SDE Backend employees
sde_backend = df[df["role_clean"] == "SDE Backend"].copy()

bins = [0, 1, 3, 5, 100]
labels = ["0-1", "2-3", "4-5", "6+"]

sde_backend["exp_band"] = pd.cut(
    sde_backend["years_exp"],
    bins=bins,
    labels=labels,
    include_lowest=True
)

In [86]:
# Calculating median CTC for each experience band (Ai Assisted)
exp_ctc = sde_backend.groupby("exp_band")["current_ctc"].median()

exp_ctc

/tmp/ipykernel_2006/4115212853.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  exp_ctc = sde_backend.groupby("exp_band")["current_ctc"].median()


,current_ctc
exp_band,
0-1,11.65
2-3,20.00
4-5,25.85
6+,40.40


In [87]:
# Calculating percentage growth between experience bands
growth_23 = ((20.00 - 11.65) / 11.65) * 100
growth_45 = ((25.85 - 20.00) / 20.00) * 100
growth_6 = ((40.40 - 25.85) / 25.85) * 100

print(f"2-3 years: {growth_23:.2f}%")
print(f"4-5 years: {growth_45:.2f}%")
print(f"6+ years: {growth_6:.2f}%")

2-3 years: 71.67%
4-5 years: 29.25%
6+ years: 56.29%


### ANSWER Q3.2

The median CTC for SDE Backend professionals increases from 11.65 LPA for 0–1 years of experience to 40.40 LPA for 6+ years. The largest salary jump (71.67%) occurs when moving from the 0–1 year band to the 2–3 year band, highlighting the value of gaining early industry experience.

#TASK 3 Q3.3

In [88]:
# Filter only SDE roles
sde = df[df["role_clean"].isin(["SDE Backend", "SDE Frontend", "SDE Full-Stack"])]

In [89]:
# Comparing median CTC of SDEs with and without AWS
with_aws = sde[sde["skills"].str.contains("AWS", na=False)]["current_ctc"].median()

without_aws = sde[~sde["skills"].str.contains("AWS", na=False)]["current_ctc"].median()

print("With AWS:", with_aws)
print("Without AWS:", without_aws)

With AWS: 20.95
Without AWS: 21.9


In [90]:
# Comparing  median CTC of SDEs with and without ML
with_ml = sde[sde["skills"].str.contains("ML", na=False)]["current_ctc"].median()

without_ml = sde[~sde["skills"].str.contains("ML", na=False)]["current_ctc"].median()

print("With ML:", with_ml)
print("Without ML:", without_ml)

With ML: 23.95
Without ML: 21.3


In [91]:
# Comparing median CTC of SDEs with and without System Design
with_sd = sde[sde["skills"].str.contains("System Design", na=False)]["current_ctc"].median()

without_sd = sde[~sde["skills"].str.contains("System Design", na=False)]["current_ctc"].median()

print("With System Design:", with_sd)
print("Without System Design:", without_sd)

With System Design: 25.3
Without System Design: 21.0


In [92]:
# Compare median CTC of SDEs with and without Kubernetes
with_k8 = sde[sde["skills"].str.contains("Kubernetes", na=False)]["current_ctc"].median()

without_k8 = sde[~sde["skills"].str.contains("Kubernetes", na=False)]["current_ctc"].median()

print("With Kubernetes:", with_k8)
print("Without Kubernetes:", without_k8)

With Kubernetes: 23.2
Without Kubernetes: 21.5


In [93]:
# Calculating skill premium percentages
aws_premium = ((with_aws - without_aws) / without_aws) * 100
ml_premium = ((with_ml - without_ml) / without_ml) * 100
sd_premium = ((with_sd - without_sd) / without_sd) * 100
k8_premium = ((with_k8 - without_k8) / without_k8) * 100

print(f"AWS: {aws_premium:.2f}%")
print(f"ML: {ml_premium:.2f}%")
print(f"System Design: {sd_premium:.2f}%")
print(f"Kubernetes: {k8_premium:.2f}%")

AWS: -4.34%
ML: 12.44%
System Design: 20.48%
Kubernetes: 7.91%


### ANSWER Q3.3

Among SDE professionals, System Design provides the highest salary premium, increasing the median CTC by 20.48% compared to those without the skill. Students aiming for higher-paying software engineering roles should prioritize learning System Design along with core development skills.

#TASK 3 Q3.4

In [94]:
# Filtering only SDE Backend employees
backend = df[df["role_clean"] == "SDE Backend"]

In [95]:
# Calculating median CTC for each company type
company_ctc = backend.groupby("company_type_clean")["current_ctc"].median()

company_ctc

,current_ctc
company_type_clean,
Early-stage,18.60
MNC,20.45
Mid-size,19.50
Unicorn,27.35


In [96]:
# Calculating percentage premium compared to Unicorn
mnc = ((27.35 - 20.45) / 27.35) * 100
mid = ((27.35 - 19.50) / 27.35) * 100
early = ((27.35 - 18.60) / 27.35) * 100

print(f"MNC: -{mnc:.2f}%")
print(f"Mid-size: -{mid:.2f}%")
print(f"Early-stage: -{early:.2f}%")

MNC: -25.23%
Mid-size: -28.70%
Early-stage: -31.99%


### ANSWER Q3.4

For SDE Backend professionals, Unicorn companies offer the highest median CTC at 27.35 LPA. This is 25.23% higher than MNCs, indicating that targeting Unicorn companies can lead to significantly better salary opportunities for backend engineers.

# TASK 3 Q3.5

In [97]:
# Creating experience bands for all employees (AI assisted)
bins = [0, 1, 3, 5, 100]
labels = ["0-1", "2-3", "4-5", "6+"]

df["exp_band"] = pd.cut(df["years_exp"], bins=bins, labels=labels, include_lowest=True)

In [98]:
# Calculatingg group median CTC based on role, company type and experience
df["group_median"] = df.groupby(
    ["role_clean", "company_type_clean", "exp_band"]
)["current_ctc"].transform("median")

/tmp/ipykernel_2006/3908850783.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df["group_median"] = df.groupby(


In [99]:
# Counting employees in each group
df["group_size"] = df.groupby(
    ["role_clean", "company_type_clean", "exp_band"]
)["employee_id"].transform("count")

/tmp/ipykernel_2006/3496162683.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df["group_size"] = df.groupby(


In [100]:
# Calculating salary gap from the group median
df["gap"] = df["current_ctc"] - df["group_median"]

In [101]:
# Displaying top 10 underpaid professionals
underpaid = df[df["group_size"] >= 10].sort_values("gap").head(10)

underpaid[[
    "employee_id",
    "role_clean",
    "company_type_clean",
    "years_exp",
    "current_ctc",
    "group_median",
    "gap"
]]

,employee_id,role_clean,company_type_clean,years_exp,current_ctc,group_median,gap
904,BLR0925,SDE Backend,MNC,4,19.4,27.30,-7.90
250,BLR0072,SDE Backend,MNC,4,19.5,27.30,-7.80
249,BLR0733,Business Analyst,MNC,7,31.1,38.50,-7.40
402,BLR0502,DevOps Engineer,Unicorn,2,17.7,24.80,-7.10
420,BLR0768,DevOps Engineer,Unicorn,2,17.8,24.80,-7.00
728,BLR0301,Product Manager,MNC,2,24.7,31.45,-6.75
386,BLR0453,Business Analyst,MNC,6,32.1,38.50,-6.40
753,BLR0611,Product Manager,MNC,2,25.1,31.45,-6.35
688,BLR0203,SDE Backend,MNC,4,21.3,27.30,-6.00
711,BLR0451,SDE Frontend,MNC,4,19.5,25.50,-6.00


### ANSWER Q3.5

The most underpaid employee in the dataset is BLR0925 SDE Backend, MNC, earning 7.90 LPA below the median salary for professionals with the same role, company type, and experience level. Such comparisons can help employees identify whether they should negotiate a salary increase or explore better-paying opportunities.

#TASK 4

In [102]:
# PrintIing the final salary report

print("="*65)
print("          BANGALORE TECH SALARY DECODER")
print("      Built by Amit Nair | The Unlox Academy| 2-Hour Live Project")
print("="*65)

print("\nDataset : 1000 Bengaluru Tech Professionals")
print("Period  : 2024 Employment Snapshot")

print("\n------ MEDIAN CTC BY ROLE (LPA) ------")
print(role_ctc["Median"])

print("\n------ SDE BACKEND CTC BY EXPERIENCE ------")
print(exp_ctc)

print("\n------ COMPANY TYPE PREMIUM (SDE Backend) ------")
print(company_ctc)

print("\n------ TOP 10 MOST UNDERPAID PROFESSIONALS ------")
print(underpaid[[
    "employee_id",
    "role_clean",
    "company_type_clean",
    "years_exp",
    "gap"
]])

print("="*65)
print("                END OF REPORT")
print("="*65)

          BANGALORE TECH SALARY DECODER
      Built by Amit Nair | The Unlox Academy| 2-Hour Live Project

Dataset : 1000 Bengaluru Tech Professionals
Period  : 2024 Employment Snapshot

------ MEDIAN CTC BY ROLE (LPA) ------
role_clean
Product Manager            31.30
ML Engineer                27.50
Data Scientist             24.15
Infrastructure Engineer    23.45
SDE Full-Stack             22.35
DevOps Engineer            22.10
SDE Frontend               21.20
SDE Backend                21.10
Business Analyst           19.90
UI/UX Designer             18.90
Data Analyst               16.90
Name: Median, dtype: float64

------ SDE BACKEND CTC BY EXPERIENCE ------
exp_band
0-1    11.65
2-3    20.00
4-5    25.85
6+     40.40
Name: current_ctc, dtype: float64

------ COMPANY TYPE PREMIUM (SDE Backend) ------
company_type_clean
Early-stage    18.60
MNC            20.45
Mid-size       19.50
Unicorn        27.35
Name: current_ctc, dtype: float64

------ TOP 10 MOST UNDERPAID PROFESSIONALS 

# TASK 5

## MARKDOWN

1. Product Managers have the highest median CTC at 31.30 LPA, while Data Analysts have the lowest median CTC at 16.90 LPA, showing a significant salary difference across roles.

2. For SDE Backend professionals, the median salary increases from 11.65 LPA (0–1 years) to 40.40 LPA (6+ years). The largest salary jump (71.67%) occurs between the 0–1 and 2–3 year experience bands.

3. Among SDEs, System Design provides the highest salary premium (20.48%) compared to professionals without the skill, making it one of the most valuable skills to learn.